src/lib/logging.ts

```diff
@@ -0,0 +1,453 @@
+                case 'debug':
+                    console.debug(msg, logEntry.objects);
+                    break;
+                case 'info':
+                    console.info(msg, logEntry.objects);
+                    break;
+                case 'warn':
+                    console.warn(msg, logEntry.objects);
+                    break;
+                case 'error':
+                    console.error(msg, logEntry.objects);
+                    break;
+                default:
+                    console.log(`{${logEntry.level}} ${msg}`, logEntry.objects);
+            }
+        });
+
+        this.consoleLoggerRegistered = true;
+        return this;
+    }
+}
+
+export const logging = new LogManager();
+
+/**
+ * Main logging library, to view the logs a logger listener must be added. The
+ * Console Logger is already implemented for this project.
+ *
+ * @export
+ * @class Logger
+ */
+export class Logger {
+    private logManager: EventEmitter2;
+    private minLevel: number;
+    private module: string;
+    private readonly levels: { [key: string]: number } = {
+        trace: 1,
+        debug: 2,
+        info: 3,
+        warn: 4,
+        error: 5,
+    };
+
+    /**
+     * Creates an instance of Logger.
+     * @param {EventEmitter2} logManager
+     * @param {string} module
+     * @param {string} minLevel
+     * @memberof Logger
+     */
+    constructor(logManager: EventEmitter2, module: string, minLevel: string) {
+        this.logManager = logManager;
+        this.module = module;
+        this.minLevel = this.levelToInt(minLevel);
+    }
+
+    /**
+     * Converts a string level (trace/debug/info/warn/error) into a number
+     *
+     * @param minLevel
+     */
+    private levelToInt(minLevel: string): number {
+        if (minLevel.toLowerCase() in this.levels)
+            return this.levels[minLevel.toLowerCase()];
+        else return 99;
+    }
+
+    /**
+     * Central logging method.
+     * @param logLevel
+     * @param message
+     */
+    public log(logLevel: string, message: string, objects?: any): void {
+        const level = this.levelToInt(logLevel);
+        if (level < this.minLevel) return;
+
+        const logEntry: LogEntry = {
+            level: logLevel,
+            module: this.module,
+            message,
+            objects,
+            traceId: undefined,
+        };
+
+        // Obtain the line/file through a thoroughly hacky method
+        // This creates a new stack trace and pulls the caller from it.  If the caller
+        // if .trace()
+        // const error = new Error('');
+        // if (error.stack) {
+        //     const cla = error.stack.split('\n');
+        //     let idx = 1;
+        //     while (idx < cla.length && cla[idx].includes('at Logger.Object.')) idx++;
+        //     if (idx < cla.length) {
+        //         logEntry.location = cla[idx].slice(cla[idx].indexOf('at ') + 3, cla[idx].length);
+        //     }
+        // }
```

Comment:
Why is this commented-out code needed? I presume it might be turned on by developers for some reason?

There is a dangling comment in line 240 'If the caller'

Reply:
Trying to get a line number for tracing, left it there to come back to later to see if there is a better way to get the line number of the original log call.